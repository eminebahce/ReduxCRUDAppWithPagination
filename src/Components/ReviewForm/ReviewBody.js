import React, {Component} from 'react'
import {TextField} from "@material-ui/core";

class ReviewBody extends Component {

    render() {
        return (
            <TextField id="formControlsTextarea"
                       rowsMax="4"
                       label="Review:"
                       placeholder="Enter your review here"
                       required={true}
                       multiline
                       data-test="formTextArea"
            />
        );
    }
}

export default ReviewBody;
