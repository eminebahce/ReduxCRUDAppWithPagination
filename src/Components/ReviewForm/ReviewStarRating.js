import React from 'react'
import StarRatingComponent from 'react-star-rating-component';
import PropTypes from 'prop-types';


/*
    react-star-rating-component
    https://github.com/voronianski/react-star-rating-component
*/

class ReviewStarRating extends React.Component {

  onStarClick = (nextValue, prevValue) => {
    this.props.rate(nextValue);
  };

  render() {
    return (
        <StarRatingComponent
            name="star-rating"
            onStarClick={this.onStarClick}
            value={this.props.value}
            data-test="star-rating"
        />
    );
  }
}
ReviewStarRating.propTypes = {
  onStarClick : PropTypes.func
}

export default ReviewStarRating