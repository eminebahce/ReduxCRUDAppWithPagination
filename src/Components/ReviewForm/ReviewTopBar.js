import React, {Component} from 'react';
import {AppBar, Toolbar, Typography, IconButton, Grid} from '@material-ui/core'
import Icon from '@material-ui/core/Icon'
import MenuIcon from '@material-ui/icons/Menu';
import {withStyles} from '@material-ui/core/styles';

const styles = {
    grow: {
        flexGrow: 1,
        marginLeft: 10,
        marginRight: 10
    },
    menuButton: {
        marginLeft: -12,
    }
};

class ReviewTopBar extends Component {
    render() {
        return (
            <div data-test='top-bar'>
                <Grid container justify="center" alignContent="center">
                    <Grid item xs={6}>
                        <AppBar color="primary" position="static" data-test='app-bar'>
                            <Toolbar>
                                <IconButton style={styles.menuButton} color="inherit" aria-label="Menu">
                                    <MenuIcon/>
                                </IconButton>
                                <Icon>star</Icon>
                                <span><Typography variant="h6" style={styles.grow}
                                                  color="inherit">Reviews</Typography></span>
                                <Icon>star</Icon>
                            </Toolbar>
                        </AppBar>
                    </Grid>
                </Grid>
            </div>
        );
    }
}

export default withStyles(styles)(ReviewTopBar);
