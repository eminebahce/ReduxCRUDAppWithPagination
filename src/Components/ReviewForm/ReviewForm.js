import React, {Component} from 'react'
import {
    Grid,
    Paper,
    FormControl,
    Fab,
    FormLabel,
    Tooltip
} from '@material-ui/core'
import AddIcon from "@material-ui/icons/Add";
import ReviewStarRating from './ReviewStarRating'
import {createReview} from '../../Actions/ReviewActions';
import {connect} from "react-redux";
import {ToastContainer, toast} from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import ReviewBody from "./ReviewBody";
import PropTypes from 'prop-types';

class ReviewForm extends Component {

    state = {
        rate: 0
    };

    onChange = (rate) => {
        this.setState({rate})
    }

    onNotification = (alert) => {
        if (alert) {
            toast.success("New review has been added successfully!", {
                position: toast.POSITION.TOP_RIGHT
            });
        } else {
            toast.error("Something went wrong!", {
                position: toast.POSITION.TOP_RIGHT
            });
        }
    }

    onSubmit = (event) => {
        event.preventDefault();
        const review = {
            'body': event.target.formControlsTextarea.value,
            'rating': this.state.rate
        }
        if (review.body.length !== 0) {
            this.props.createReview(review, this.calculatePage(), this.onNotification);
            this.resetForm(event);
        }
    }

    calculatePage = () => {
        return Math.ceil((parseInt(this.props.totalCount) + 1) / 5);
    }

    resetForm = (event) => {
        event.target.formControlsTextarea.value = '';
        this.onChange(0);
    }

    render() {
        return (
            <div data-test="review-form">
                <Grid container justify="center" alignContent="center">
                    <Grid item xs={6}>
                        <Paper elevation={3}>
                            <form style={{width: "100%", height: "100%"}} onSubmit={this.onSubmit}>
                                <FormControl style={{marginLeft: '10px', width: "90%", marginTop: '25px'}}
                                             required={true}>
                                    <ReviewBody/>
                                </FormControl>
                                <FormControl style={{marginLeft: '10px', marginTop: '15px'}} required={true} fullWidth>
                                    <FormLabel>Rating</FormLabel>
                                </FormControl>
                                <FormControl style={{marginLeft: '10px', marginTop: '10px'}}>
                                        <ReviewStarRating rate={this.onChange} value={this.state.rate}/>
                                </FormControl>
                                <FormControl fullWidth>
                                    <Tooltip title="Click to add the review">
                                        <Fab type="submit" variant="round" color="primary" size="medium"
                                             style={{marginTop: '20px', marginBottom: '20px', marginLeft: '10px'}}>
                                            <AddIcon/>
                                        </Fab>
                                    </Tooltip>
                                </FormControl>
                            </form>
                        </Paper>
                    </Grid>
                </Grid>
                <ToastContainer/>
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    reviewSize: state.reviewSize,
    totalCount: state.totalCount
});

ReviewForm.propTypes = {
    onSubmit : PropTypes.func,
    onChange: PropTypes.func,
    onNotification: PropTypes.func,
    calculatePage:PropTypes.func,
    resetForm:PropTypes.func
};

export default connect(mapStateToProps, {createReview})(ReviewForm);