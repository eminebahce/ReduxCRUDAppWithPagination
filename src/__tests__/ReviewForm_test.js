import React from 'react';
import ReviewForm from '../../src/Components/ReviewForm/ReviewForm';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import store from '../store';
import ReviewStarRating from '../../src/Components/ReviewForm/ReviewStarRating'
import ReviewBody from '../../src/Components/ReviewForm/ReviewBody'
import { shallow } from 'enzyme';
import {Button, Form} from "react-bootstrap";
import checkPropTypes from "check-prop-types";

describe('<ReviewForm /> component', () => {
    beforeEach(() => {
        const props = {
            onSubmit: () => {},
            onChange: () => {},
            onNotification: () => {},
            calculatePage: () => {},
            resetForm: () => {}
        }
    });
    const checkProps = (component, expectedProps) => {
        const propsErr = checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
        return propsErr;
    };
    describe('Checking propTypes', () => {
        it('should not throw a warning', () => {
            const expectedProps = {
                onSubmit: () => {},
                onChange: () => {},
                onNotification: () => {},
                calculatePage: () => {},
                resetForm: () => {}
            };
            const propsError = checkProps(ReviewForm, expectedProps);
            expect(propsError).toBeUndefined();
        });
    });
    it('should render without error', () => {
        const component = shallow(<ReviewForm />);
        expect(component.find(`[data-test='review-form']`)).toBeTruthy();
    });
    it('should return form body, star element and save button', () => {
        const component = renderer.create(
            <Provider store={store}>
                <Form>
                    <ReviewBody/>
                    <ReviewStarRating />
                    <Button id="button" bsStyle="primary" bsSize="small" type="submit"></Button>
                </Form>
            </Provider>
        );
       let tree = component.toJSON();
       expect(tree).toMatchSnapshot();
    });
});

