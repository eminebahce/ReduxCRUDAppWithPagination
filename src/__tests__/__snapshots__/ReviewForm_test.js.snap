// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReviewForm /> component should return form body, star element and save button 1`] = `
<form
  className=""
>
  <div
    className="MuiFormControl-root-1"
    data-test="formTextArea"
  >
    <label
      className="MuiFormLabel-root-16 MuiFormLabel-required-21 MuiInputLabel-required-9 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
      data-shrink={false}
      htmlFor="formControlsTextarea"
    >
      Review:
      <span
        className="MuiFormLabel-asterisk-22"
      >
         *
      </span>
    </label>
    <div
      className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-formControl-37 MuiInput-formControl-24 MuiInputBase-multiline-44 MuiInput-multiline-29"
      onClick={[Function]}
    >
      <div
        className="MuiPrivateTextarea-root-53"
      >
        <textarea
          aria-hidden="true"
          className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
          readOnly={true}
          rows="1"
          tabIndex={-1}
          value=""
        />
        <textarea
          aria-hidden="true"
          className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
          readOnly={true}
          rows={1}
          tabIndex={-1}
        />
        <textarea
          aria-invalid={false}
          className="MuiPrivateTextarea-textarea-54 MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputMultiline-48 MuiInput-inputMultiline-33"
          disabled={false}
          id="formControlsTextarea"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Enter your review here"
          required={true}
          rows={1}
          style={
            Object {
              "height": 19,
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="dv-star-rating"
    style={
      Object {
        "display": "inline-block",
        "position": "relative",
      }
    }
  >
    <input
      checked={false}
      className="dv-star-rating-input"
      id="star-rating_5"
      name="star-rating"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
          "marginLeft": -9999,
          "position": "absolute",
        }
      }
      type="radio"
      value={5}
    />
    <label
      className="dv-star-rating-star dv-star-rating-empty-star"
      htmlFor="star-rating_5"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "color": "#333",
          "cursor": "pointer",
          "float": "right",
        }
      }
    >
      <i
        style={
          Object {
            "fontStyle": "normal",
          }
        }
      >
        ★
      </i>
    </label>
    <input
      checked={false}
      className="dv-star-rating-input"
      id="star-rating_4"
      name="star-rating"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
          "marginLeft": -9999,
          "position": "absolute",
        }
      }
      type="radio"
      value={4}
    />
    <label
      className="dv-star-rating-star dv-star-rating-empty-star"
      htmlFor="star-rating_4"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "color": "#333",
          "cursor": "pointer",
          "float": "right",
        }
      }
    >
      <i
        style={
          Object {
            "fontStyle": "normal",
          }
        }
      >
        ★
      </i>
    </label>
    <input
      checked={false}
      className="dv-star-rating-input"
      id="star-rating_3"
      name="star-rating"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
          "marginLeft": -9999,
          "position": "absolute",
        }
      }
      type="radio"
      value={3}
    />
    <label
      className="dv-star-rating-star dv-star-rating-empty-star"
      htmlFor="star-rating_3"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "color": "#333",
          "cursor": "pointer",
          "float": "right",
        }
      }
    >
      <i
        style={
          Object {
            "fontStyle": "normal",
          }
        }
      >
        ★
      </i>
    </label>
    <input
      checked={false}
      className="dv-star-rating-input"
      id="star-rating_2"
      name="star-rating"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
          "marginLeft": -9999,
          "position": "absolute",
        }
      }
      type="radio"
      value={2}
    />
    <label
      className="dv-star-rating-star dv-star-rating-empty-star"
      htmlFor="star-rating_2"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "color": "#333",
          "cursor": "pointer",
          "float": "right",
        }
      }
    >
      <i
        style={
          Object {
            "fontStyle": "normal",
          }
        }
      >
        ★
      </i>
    </label>
    <input
      checked={false}
      className="dv-star-rating-input"
      id="star-rating_1"
      name="star-rating"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
          "marginLeft": -9999,
          "position": "absolute",
        }
      }
      type="radio"
      value={1}
    />
    <label
      className="dv-star-rating-star dv-star-rating-empty-star"
      htmlFor="star-rating_1"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "color": "#333",
          "cursor": "pointer",
          "float": "right",
        }
      }
    >
      <i
        style={
          Object {
            "fontStyle": "normal",
          }
        }
      >
        ★
      </i>
    </label>
  </div>
  <button
    className="btn btn-sm btn-primary"
    disabled={false}
    id="button"
    type="submit"
  />
</form>
`;
